dataavailability/dataavailability.go

```diff
@@ -55,60 +60,93 @@ func (d *DataAvailability) PostSequence(ctx context.Context, sequences []types.S
 	return d.backend.PostSequence(ctx, batchesData)
 }
 
-// GetBatchL2Data tries to return the data from a batch, in the following priorities
+// GetBatchL2Data tries to return the data from a batch, in the following priorities. batchNums should not include forced batches.
 // 1. From local DB
-// 2. From Sequencer
+// 2. From Trusted Sequencer (if not self)
 // 3. From DA backend
 func (d *DataAvailability) GetBatchL2Data(batchNums []uint64, batchHashes []common.Hash, dataAvailabilityMessage []byte) ([][]byte, error) {
 	if len(batchNums) != len(batchHashes) {
-		return nil, fmt.Errorf("invalid L2 batch data retrieval arguments, %d != %d", len(batchNums), len(batchHashes))
+		return nil, fmt.Errorf(invalidBatchRetrievalArgs, len(batchNums), len(batchHashes))
+	}
+	localData, err := d.state.GetBatchL2DataByNumbers(d.ctx, batchNums, nil)
+	if err != nil {
+		return nil, err
 	}
 
-	data, err := d.localData(batchNums, batchHashes)
-	if err == nil {
+	data, err := checkBatches(batchNums, batchHashes, localData)
+	if err != nil {
+		log.Warnf(failedDataRetrievalTemplate, batchNums, err.Error())
+	} else {
 		return data, nil
 	}
 
 	if !d.isTrustedSequencer {
-		data, err = d.trustedSequencerData(batchNums, batchHashes)
+		data, err = d.rpcData(batchNums, batchHashes, d.zkEVMClient.BatchesByNumbers)
 		if err != nil {
-			log.Warnf("trusted sequencer failed to return data for batches %v: %s", batchNums, err.Error())
+			log.Warnf(failedDataRetrievalTemplate, batchNums, err.Error())
 		} else {
 			return data, nil
 		}
 	}
-
 	return d.backend.GetSequence(d.ctx, batchHashes, dataAvailabilityMessage)
 }
 
-// localData retrieves batches from local database and returns an error unless all are found
-func (d *DataAvailability) localData(numbers []uint64, hashes []common.Hash) ([][]byte, error) {
-	data, err := d.state.GetBatchL2DataByNumbers(d.ctx, numbers, nil)
+// GetForcedBatchL2Data retrieves, checks, and returns the raw data associated with forced batches.
+// 1. From local DB
+// 2. From Trusted Sequencer (if not self)
+func (d *DataAvailability) GetForcedBatchL2Data(batchNums []uint64, batchHashes []common.Hash) ([][]byte, error) {
```

Comment:
No need to look for forced batch data from the trusted sequencer as we will always have this on our DB (it's synced from calldata in the contract)